### ISSUE TITLE
VerticalCAS fails to reproduce MDP results when training

### ISSUE THREAD

**Issue by nskh** [sisl/VerticalCAS#5]
I'm attempting to retrain VerticalCAS neural networks but my training is proving to be remarkably unstable. I've been working with the VerticalCAS repo for a few months and in the last few weeks have been unable to duplicate the behavior of the MDP (image below, generated by my own code since I was unable to get PolicyViz to work as documented in https://github.com/sisl/VerticalCAS/issues/3)

![Figure 13](https://github.com/sisl/VerticalCAS/assets/8115839/3f61f585-d72d-463f-9bec-029951955ab9)

I have, in the past, been able to successfully retrain an approximation of the MDP using a neural network using (relatively) standard VerticalCAS code (photo below). This image was the policy after only one epoch, and I recall training this behavior to have been fairly reliable (happened > 50% of the time I tried to train). Additionally, losses were small and accuracy was high from early on in training: one value I saw was `loss: 0.0463 - accuracy: 0.9569`
![Screenshot 2023-10-17 at 12 17 33 PM](https://github.com/sisl/VerticalCAS/assets/8115839/6e76e04e-92e9-4288-bba6-2ff8dfd3a41e)

I've since been unable to reproduce this behavior and get nonsense policies with terrible loss and accuracy (`loss: 61075845120.0000 - accuracy: 0.5324`). I think these massive losses indicate some sort of instability when training, but I haven't seen them until a few weeks ago. This happens every time I try to train.
![Screenshot 2023-11-09 at 5 06 58 PM](https://github.com/sisl/VerticalCAS/assets/8115839/1306f427-e422-40d7-a642-6ce3361756bc)

I did have to make some modifications to the code to get it to run on modern Python (3.10.12) and Tensorflow (2.14.0), detailed below. Note we're both using M1 MacBook Pros, but again this wasn't a problem a few months ago.

- changing some indexing in `GenTrainingData.py` to `//` from `/` for integer division
- using the `AsymMSE` from the [`dasc` branch](https://github.com/sisl/VerticalCAS/tree/dasc) or [HorizontalCAS repo](https://github.com/sisl/HorizontalCAS) (they appear to be the same) so the Theano dependency goes away
- removing `init='uniform'` when initializing `Dense` layers

However, I had made all these changes a few months ago when I was able to successfully retrain, and now nothing I've tried seems to work. Another member of my lab has also tried to retrain these networks and has had the same issues as I have.

Do you have any suggested resolution to these issues? And do others experience these sorts of issues when attempting to retrain?

**Comment by smkatz12** [comment]
We have not made any changes to the code in the last couple months, so it is difficult to know what might be causing the issue. Maybe look back at any changes you have made since it was last working for you?

**Comment by nskh** [comment]
I tried reverting to past commits that performed well before and got the same issues. The next thing I tried was trying to use a fresh clone of this VerticalCAS repo without any of my own changes for verification in the loop, but I still ran into the same problems. I had to make a few changes to modernize the code to run on Tensorflow and Python3, as documented above, but they're very minimal. Do you have a working instance of VerticalCAS running on your end? Can you share some config and/or environment details if so?

**Comment by stschaef** [comment]
I'm on `tf == 2.13.0` and `keras == 2.13.1` on an M1 Macbook, ran in a Jupyter notebook.

Yeah, I think some config/environment issue may clear things up

Running the same network as in you code, same `asymMSE`, same network architecture, activations all `ReLU`, weight initialization uniform, etc, we learn a trivial policy. With the minor caveat that I cannot initialize weights in the same manner that you do, in my current version of Keras, I must use `kernel_initializer="uniform"` rather than `init="uniform"`. However, this shouldn't matter too much (but maybe it does :( )

From what I can gather, it stems from a dying `ReLU`. When I swap activations out from `ReLU` to a `LeakyReLU`, we learn the correct behavior. This is nice because we can train the proper policy now, but we are interested in investigating a notion of network repair at training time, and this process is sensitive to which activation is used. To this end, it would be ideal if we could replicate the version of your network that only uses `ReLU` activations.

I've tried playing around with different schemes of weight initialization with no luck thus far. If you have any ideas on what's going on here I'd love to hear them. In any case, if you have the time to dump your configuration that successfully ran the network we should be able to figure out the rest. 

Thanks
